Classes/common/touchdb/TD_Database.m

```diff
@@ -438,24 +577,43 @@ - (BOOL)close
     _activeReplicators = nil;
 
     [_fmdbQueue close];
+    _fmdbQueue = nil;
+    _keyProviderToOpenDB = nil;
 
     _open = NO;
     _transactionLevel = 0;
+    
     return YES;
 }
 
-- (BOOL)deleteDatabase:(NSError**)outError
+- (BOOL)deleteDatabase:(NSError **)outError
 {
-    CDTLogInfo(CDTDATASTORE_LOG_CONTEXT, @"Deleting %@", _path);
     [[NSNotificationCenter defaultCenter] postNotificationName:TD_DatabaseWillBeDeletedNotification
                                                         object:self];
-    if (_open) {
-        if (![self close]) return NO;
-    } else if (!self.exists) {
-        return YES;
+
+    if (_open && ![self close]) {
+        CDTLogError(CDTDATASTORE_LOG_CONTEXT, @"Database at path %@ could not be closed", _path);
+
+        return NO;
+    }
+
+    return [[self class] deleteClosedDatabaseAtPath:_path error:outError];
+}
+
++ (BOOL)deleteClosedDatabaseAtPath:(NSString *)path error:(NSError **)outError
+{
+    CDTLogInfo(CDTDATASTORE_LOG_CONTEXT, @"Deleting %@", path);
+    
+    BOOL success = YES;
+
+    if ([TD_Database existsDatabaseAtPath:path]) {
+        NSString *attachmentsPath = [TD_Database attachmentStorePathWithDatabasePath:path];
+
+        success =
+            (removeItemIfExists(path, outError) && removeItemIfExists(attachmentsPath, outError));
```

Comment:
Personally, I'm not sure the `if` check is needed, given that the method name `removeItemIfExists` sounds like it'll do something similar anyway, and anyway something could easily delete the database between the `existsDatabaseAtPath:path` call and the `removeItemIfExists` call it appears to be attempting to protect.

It's important to remember that a lot of this code assumed single threading. While I corrected most when I moved to GCD, we should probably be cleaning up code like this that relies on those assumptions as we go.